third_party/move/move-compiler-v2/src/inliner.rs

```diff
@@ -0,0 +1,1284 @@
+pub fn run_inlining(env: &mut GlobalEnv) {
+    // Get non-inline function roots for running inlining.
+    // Also generate an error for any target inline functions lacking a body to inline.
+    let mut todo = get_targets(env);
+
+    // Only look for inlining sites if we have targets to inline into.
+    if !todo.is_empty() {
+        // Recursively find callees of each target with a function body.
+
+        // The call graph reachable from targets, represented by a map from each function to the set
+        // of functions it calls.  The domain is limited to functions with function bodies.
+        let mut call_graph: BTreeMap<QualifiedFunId, BTreeSet<QualifiedFunId>> = BTreeMap::new();
+
+        // For each function `caller` calling an inline function `callee`, we record the set of all
+        // call sites where `caller` calls `callee` (for error messages).
+        let mut inline_function_call_site_locations: CallSiteLocations = CallSiteLocations::new();
+
+        // Update call_graph and inline_function_call_site_locations for all reachable calls.
+        let mut visited_functions = BTreeSet::new();
+        while let Some(id) = todo.pop_first() {
+            if visited_functions.insert(id) {
+                if let Some(def) = env.get_function(id).get_def().deref() {
+                    let callees_with_sites = def.called_funs_with_callsites();
+                    for (callee, sites) in callees_with_sites {
+                        todo.insert(callee);
+                        call_graph.entry(id).or_default().insert(callee);
+                        if env.get_function(callee).is_inline() {
+                            inline_function_call_site_locations.insert((id, callee), sites);
+                        }
+                    }
+                }
+            }
+        }
+
+        // Get a list of all reachable functions calling inline functions, in bottom-up order.
+        // If there are any cycles, this call displays an error to the user and returns None.
+        if let Ok(functions_needing_inlining) = functions_needing_inlining_in_order(
+            env,
+            &call_graph,
+            inline_function_call_site_locations,
+        ) {
+            // We inline functions bottom-up, so that any inline function which itself has calls to
+            // inline functions has already had its stuff inlined.
+            let mut inliner = Inliner::new(env);
+            for fid in functions_needing_inlining.iter() {
+                inliner.do_inlining_in(*fid);
+            }
+
+            // Now that all inlining finished, actually update function bodies in env.
+            for (fun_id, funexpr_after_inlining) in inliner.funexprs_after_inlining {
+                if let Some(changed_funexpr) = funexpr_after_inlining {
+                    let oldexp = env.get_function(fun_id);
+                    // TODO/BUGBUG: Is this the right thing to do here?
```

Comment:
Wolfgang had some concern about whether only recording changed function bodies for functions which are targets is a problem with "whole program", but he seems to have dropped it himself, so I'll delete this comment.

Reply:
... and after merging in @wrwg 's recent push, I see what he meant.  It seems that I need to be sure to inline into any functions reachable from targets, even if they themselves are not in a target module.  Otherwise I see persistent calls to inline functions in functions outside of the target modules, but reachable from the target modules.  

Hopefully that's sufficient, and we don't have to do inlining on all libraries for every compilation.